File: projecteuler/problem59/c/xordecryption.c
#include <stdio.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

unsigned int gl_cleanbytes;

/*
 *function to print a single character
*/
void my_char(char c)
{
  write(1,&c,1);
}

/*
 *function to print a null-terminated string (char *)
*/
void my_str(char *s)
{
  if (s != NULL)
    for(; *s!='\0'; ++s)
      my_char(*s);
}

/*
 *pre: takes a key, ciphertext, and the number of bytes in the ciphertext
 *post: performs a repeating-key xor decryption, returns the result
*/
char *xordecrypt(char *key, unsigned int nbytes, char *ctxt)
{
  unsigned int i;
  unsigned int keylen;
  char *decrypt = (char *)malloc((nbytes+1)*sizeof(char));
  keylen = strlen(key);

  for (i=0; i<nbytes; ++i)
    decrypt[i] = ctxt[i] ^ key[i%keylen];
  decrypt[nbytes] = '\0';
  return decrypt; 
} 

/*
 *pre: takes a string (char *), percentage english (A-Za-z) threshold, and the number of bytes in the string.
 *post: returns 1 (true) or 0 based on whether the string meets the criteria and whether it is greater or less than threshold
*/
char checkenglish(char *asciilist, unsigned int threshold, unsigned int nbytes)
{
  unsigned int divisor = nbytes;
  unsigned int loopmax = nbytes;
  unsigned int lettercount;
  unsigned int i;
  
  for (lettercount=0,i=0; i<loopmax; ++i)
    if ( (asciilist[i] >= 97 && asciilist[i] <= 122) || (asciilist[i] >= 65 && asciilist[i] <= 90) || (asciilist[i] == 32) )
      ++lettercount;

  if ( ( (lettercount / (double)(divisor)) * 100) > threshold )
    return 1;
  else
    return 0;
}

/*
 *pre: takes a char * with commas separating numbers, and the size (bytes) of the char *
 *post: returns a clean char* without any commas (commas in *csv signify where to start a new index)
*/
char *parsecsv(char *csv, unsigned int nbytes)
{
  char *ret = (char *)malloc((nbytes+1)*sizeof(char));
  unsigned int i;
  unsigned int j;
  unsigned int cleancount;
  cleancount = 0;
  
  for (i=0; i<nbytes; i++)
    {
      ret[cleancount] = 0;
      j = 0;
      if (csv[i] != ',')
	{
	  for (j=i; csv[j] != ','; j++)
	    {
	      if (csv[j] >= '0' && csv[j]<='9')
		{
		  ret[cleancount] = 10*ret[cleancount];
		  ret[cleancount] += (csv[j]-48);
		}
	      else
		break;
	    }
	  ++cleancount;
	  i = j;
	}
    }
  gl_cleanbytes = cleancount;
  return ret;
}

/*
 *pre: takes a null-terminated string (char*)
 *post: returns the sum of all of the ascii vals of the individual chars up to the first '\0'
*/
unsigned int addasciivals(char *s)
{
  unsigned int sum;
  for (sum=0; *s != '\0'; ++s)
    sum += *s;
  return sum;
}

/*
 *driver and file reader function
*/
int main()
{
  FILE *file = fopen("../cipher.txt", "r");
  int n = 0;
  int c;
  char i;
  char j;
  char k;
  char *inputcsv = (char *)malloc(4096*sizeof(char));
  char *cleanctxt;
  char *key = (char *)malloc(4*sizeof(char));
  char *dec;
  
  while ((c = fgetc(file)) != EOF)
    inputcsv[n++] = (char)c;
  if (n<0)
    exit(1);
  fclose(file);

  cleanctxt = parsecsv(inputcsv,n);
  
  key[3] = '\0';

  /*try all combinations of aaa - zzz*/
  for (i=95; i<123; ++i)
    {
      key[0] = i;
      for (j=95; j<123; ++j)
	{
	  key[1] = j;
	  for (k=95; k<123; ++k)
	    {
	      key[2] = k;
	      /*if found, print relevant stats on the decrypted text*/
	      if ( checkenglish(xordecrypt(key,gl_cleanbytes-1,cleanctxt), 95, gl_cleanbytes) )
		{
		  dec = xordecrypt(key,gl_cleanbytes,cleanctxt);
		  my_str("Key: "); my_str(key); my_char('\n');
		  my_str("Sum of ASCII Values: "); printf("%d\n",(addasciivals(dec)));
		  my_str(dec); my_char('\n');
		  return 0;
		}
	    }
	}
    }
  return 1;
}

File: projecteuler/problem59/java/XORDecryption.java
import java.util.*;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

class XORDecryption
{
    /*
     *pre: takes a key, and the ciphertext
     *post: performs a repeating-key xor decryption, returns the result
     */
    public static char[] xordecrypt(char[] key, char[] ctxt)
    {
	char[] decrypt = new char[ctxt.length];

	for (int i=0; i<ctxt.length; ++i)
	    decrypt[i] = (char)(ctxt[i] ^ key[i%key.length]);
	return decrypt;
    }
    /*
     *pre: takes a char array, and percentage english (A-Za-z) threshold
     *post: returns true or false based on whether the string meets the criteria and whether it is greater or less than threshold
     */
    public static boolean checkenglish(char[] chars, int threshold)
    {
	double divisor = (double)chars.length;
	int loopmax = chars.length;
	int lettercount;
	int i;

	for (lettercount=0,i=0; i<loopmax; ++i)
	    if ( (chars[i] >= 97 && chars[i] <= 122) || (chars[i] >= 65 && chars[i] <= 90) || (chars[i] == 32) )
		++lettercount;
	
	if  ( ((lettercount / divisor) * 100) > threshold )
	    return true;
	else
	    return false;
	
    }

    /*
     *pre: takes a char[] array of ascii values
     *post: returns the sum of all of the ascii vals of the individual chars
     */
    public static int addasciivals(char[] chars)
    {
	int sum = 0;
	for (int x=0; x<chars.length; ++x)
	    sum += (int)chars[x];
	return sum;
    }
    
    /*
     *driver, file reader, and file parser function
     */
    public static void main(String[] args)
    {
	String rawin = "";
	String rawconcat = "";
	char[] key = {'a','a','a'};
	char a;
	char b;
	char c;

	try
	    {
		File fh = new File("../cipher.txt");
		BufferedReader in = new BufferedReader(new FileReader(fh));
		while ( (rawin = in.readLine()) != null )
		    rawconcat += rawin;
		in.close();
		
	    }
	catch (IOException e)
	    {
		System.out.println("error reading file");
	    }

	List<String> enclist = Arrays.asList(rawconcat.split(","));
	String[] encarr = enclist.toArray(new String[enclist.size()]);
	char[] encchars = new char[encarr.length];
	char[] dec;
	
	for (int i = 0; i < encarr.length; ++i)
	    encchars[i] = (char)Integer.parseInt(encarr[i]);

	/*try all combinations of aaa - zzz*/
	for (a=95; a<123; ++a)
	    {
		key[0] = a;
		for (b=95; b<123; ++b)
		    {
			key[1] = b;
			for (c=95; c<123; ++c)
			    {
				key[2] = c;
				/*if found, print relevant stats on the decrypted text*/
				if ( checkenglish( xordecrypt(key,encchars), 95 ) )
				    {
					dec = xordecrypt(key,encchars);
					System.out.print("Key: ");
					System.out.println(String.valueOf(key));
					System.out.print("Sum of ASCII Values: ");
					System.out.println(addasciivals(dec));
					System.out.println(String.valueOf(dec));
				    }
			    }
			
		    }
	    }
	
    }
}
